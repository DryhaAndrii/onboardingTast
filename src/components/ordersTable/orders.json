[
  { "id": 1, "name": "John Doe", "price": 120.5, "status": "pending" },
  { "id": 2, "name": "Jane Smith", "price": 89.99, "status": "shipped" },
  { "id": 3, "name": "Bob Johnson", "price": 45.0, "status": "cancelled" },
  { "id": 4, "name": "Alice Brown", "price": 77.25, "status": "pending" },
  { "id": 5, "name": "Charlie Black", "price": 150.0, "status": "shipped" },
  { "id": 6, "name": "Diana Green", "price": 55.75, "status": "cancelled" },
  { "id": 7, "name": "Eve White", "price": 95.0, "status": "pending" },
  { "id": 8, "name": "Frank Blue", "price": 65.25, "status": "shipped" },
  { "id": 9, "name": "Grace Yellow", "price": 130.0, "status": "cancelled" },
  { "id": 10, "name": "Heidi Purple", "price": 85.5, "status": "pending" },
  { "id": 11, "name": "Ivan Red", "price": 115.0, "status": "shipped" },
  { "id": 12, "name": "Judy Pink", "price": 70.75, "status": "cancelled" },
  { "id": 13, "name": "Kevin Gray", "price": 105.0, "status": "pending" },
  { "id": 14, "name": "Laura Silver", "price": 60.5, "status": "shipped" },
  { "id": 15, "name": "Mallory Gold", "price": 125.25, "status": "cancelled" },
  { "id": 16, "name": "Nancy Bronze", "price": 80.0, "status": "pending" },
  { "id": 17, "name": "Oscar Copper", "price": 135.5, "status": "shipped" },
  { "id": 18, "name": "Peggy Steel", "price": 50.25, "status": "cancelled" },
  { "id": 19, "name": "Quentin Iron", "price": 145.0, "status": "pending" },
  { "id": 20, "name": "Ruth Platinum", "price": 75.5, "status": "shipped" },
  { "id": 21, "name": "Sam Diamond", "price": 110.0, "status": "cancelled" },
  { "id": 22, "name": "Trudy Emerald", "price": 100.25, "status": "pending" },
  { "id": 23, "name": "Uma Ruby", "price": 90.0, "status": "shipped" },
  {
    "id": 24,
    "name": "Vic Sap",
    "price": 140.75,
    "status": "cancelled"
  },
  { "id": 25, "name": "Walter Topaz", "price": 55.0, "status": "pending" },
  { "id": 26, "name": "Xena Jade", "price": 115.5, "status": "shipped" },
  { "id": 27, "name": "Yolanda Quartz", "price": 65.0, "status": "cancelled" },
  { "id": 28, "name": "Zane Opal", "price": 120.25, "status": "pending" },
  { "id": 29, "name": "Aiden Pearl", "price": 85.0, "status": "shipped" },
  { "id": 30, "name": "Becca Amethyst", "price": 95.5, "status": "cancelled" }
]
