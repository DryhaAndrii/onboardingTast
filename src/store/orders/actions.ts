export const FETCH_ORDERS_REQUEST = "orders/FETCH_ORDERS_REQUEST";
export const FETCH_ORDERS_SUCCESS = "orders/FETCH_ORDERS_SUCCESS";
export const FETCH_ORDERS_FAILURE = "orders/FETCH_ORDERS_FAILURE";

export const fetchOrdersRequest = () => ({ type: FETCH_ORDERS_REQUEST });
export const fetchOrdersSuccess = (orders: any[]) => ({
  type: FETCH_ORDERS_SUCCESS,
  payload: orders,
});
export const fetchOrdersFailure = (error: string) => ({
  type: FETCH_ORDERS_FAILURE,
  payload: error,
});
