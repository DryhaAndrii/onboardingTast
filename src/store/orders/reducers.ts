import {
  FETCH_ORDERS_REQUEST,
  FETCH_ORDERS_SUCCESS,
  FETCH_ORDERS_FAILURE,
} from "./actions";
import { OrdersState } from "./types";

const initialState: OrdersState = {
  orders: [],
  loading: false,
  error: null,
};

export default function ordersReducer(
  state = initialState,
  action: any
): OrdersState {
  switch (action.type) {
    case FETCH_ORDERS_REQUEST:
      return { ...state, loading: true, error: null };
    case FETCH_ORDERS_SUCCESS:
      return { ...state, loading: false, orders: action.payload };
    case FETCH_ORDERS_FAILURE:
      return { ...state, loading: false, error: action.payload };
    default:
      return state;
  }
}
