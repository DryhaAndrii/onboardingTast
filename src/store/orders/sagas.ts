import { call, put, takeLatest } from "redux-saga/effects";
import {
  FETCH_ORDERS_REQUEST,
  fetchOrdersSuccess,
  fetchOrdersFailure,
} from "./actions";

async function fetchOrdersApi() {
  const res = await fetch("https://683d69ca199a0039e9e55d2c.mockapi.io/orders");
  return await res.json();
}

function* fetchOrdersSaga(): Generator<any, void, any> {
  try {
    const data = yield call(fetchOrdersApi);
    yield put(fetchOrdersSuccess(data));
  } catch (error: any) {
    yield put(fetchOrdersFailure(error.message));
  }
}

export function* ordersRootSaga() {
  yield takeLatest(FETCH_ORDERS_REQUEST, fetchOrdersSaga);
}
