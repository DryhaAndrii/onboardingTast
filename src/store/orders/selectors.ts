import { RootState } from "../index";

export const selectOrders = (state: RootState) => state.orders.orders;
export const selectLoading = (state: RootState) => state.orders.loading;
export const selectError = (state: RootState) => state.orders.error;
