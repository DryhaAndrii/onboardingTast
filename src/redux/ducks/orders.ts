import { call, put, takeLatest } from "redux-saga/effects";
import { RootState } from "../rootReducer";
import { Order } from "../../types/Order";

export const FETCH_ORDERS_REQUEST = "orders/FETCH_ORDERS_REQUEST";
export const FETCH_ORDERS_SUCCESS = "orders/FETCH_ORDERS_SUCCESS";
export const FETCH_ORDERS_FAILURE = "orders/FETCH_ORDERS_FAILURE";

interface OrdersState {
  data: Order[];
  loading: boolean;
  error: string | null;
}

const initialState: OrdersState = {
  data: [],
  loading: false,
  error: null,
};

export default function ordersReducer(
  state = initialState,
  action: any
): OrdersState {
  switch (action.type) {
    case FETCH_ORDERS_REQUEST:
      return { ...state, loading: true, error: null };
    case FETCH_ORDERS_SUCCESS:
      return { ...state, loading: false, data: action.payload };
    case FETCH_ORDERS_FAILURE:
      return { ...state, loading: false, error: action.payload };
    default:
      return state;
  }
}

export const fetchOrdersRequest = () => ({ type: FETCH_ORDERS_REQUEST });

export const fetchOrdersSuccess = (orders: Order[]) => ({
  type: FETCH_ORDERS_SUCCESS,
  payload: orders,
});
export const fetchOrdersFailure = (error: string) => ({
  type: FETCH_ORDERS_FAILURE,
  payload: error,
});

async function fetchOrdersApi() {
  const res = await fetch("https://683d69ca199a0039e9e55d2c.mockapi.io/orders");

  if (!res.ok) {
    const message = await res.json();
    throw new Error(`HTTP ${res.status}: ${message}`);
  }

  return await res.json();
}

function* fetchOrdersSaga() {
  try {
    const orders: Order[] = yield call(fetchOrdersApi);
    yield put(fetchOrdersSuccess(orders));
  } catch (error: any) {
    yield put(fetchOrdersFailure(error.message));
  }
}

export function* ordersSaga() {
  yield takeLatest(FETCH_ORDERS_REQUEST, fetchOrdersSaga);
}

export const selectOrderById = (id: string) => (state: RootState) =>
  state.orders.data.find((order) => order.id === id);
