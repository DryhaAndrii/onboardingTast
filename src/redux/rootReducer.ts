import { configureStore } from "@reduxjs/toolkit";
import createSagaMiddleware from "redux-saga";
import ordersReducer from "./ducks/orders";
import { useDispatch, useSelector, TypedUseSelectorHook } from "react-redux";
import { ordersSaga } from "./ducks/orders";

const sagaMiddleware = createSagaMiddleware();

export const store = configureStore({
  reducer: {
    orders: ordersReducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({ thunk: false }).concat(sagaMiddleware),
});

sagaMiddleware.run(function* () {
  yield ordersSaga();
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
