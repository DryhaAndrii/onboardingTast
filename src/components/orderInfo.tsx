import { useNavigate } from "react-router-dom";

import { ContainerOrderInfoStyled } from "../styled/styledOrderInfo";

import { selectOrderById } from "../redux/ducks/orders";
import { useAppSelector } from "../redux/rootReducer";

export function OrderInfo() {
  const params = new URLSearchParams(window.location.search);
  const id = params.get("id");
  const order = useAppSelector(selectOrderById(id || ""));
  const navigate = useNavigate();

  const buttonHandler = () => {
    navigate("/");
  };

  const getFormattedDate = (date: string) => {
    return new Intl.DateTimeFormat("eu", {
      year: "numeric",
      month: "2-digit",
      day: "2-digit",
      hour: "2-digit",
      minute: "2-digit",
    }).format(new Date(date));
  };

  return (
    <ContainerOrderInfoStyled>
      <div>
        {order && (
          <>
            <p>Client: {order.name}</p>
            <p>Date: {getFormattedDate(order.date)}</p>
            <p>Price: {order.price}</p>
            <p>Status: {order.status}</p>
            <p>Info: {order.info}</p>
          </>
        )}

        <button onClick={buttonHandler}>Go back</button>
      </div>
    </ContainerOrderInfoStyled>
  );
}
