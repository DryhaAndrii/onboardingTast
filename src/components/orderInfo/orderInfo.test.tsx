import { render, screen } from "@testing-library/react";
import { useNavigate } from "react-router-dom";

import { OrderInfo } from "./orderInfo";

import { selectOrderById } from "../../redux/ducks/orders";
import { useAppSelector } from "../../redux/rootReducer";

jest.mock("react-router-dom", () => ({
  __esModule: true,
  useNavigate: jest.fn(),
}));

jest.mock("../../redux/rootReducer", () => ({
  __esModule: true,
  useAppSelector: jest.fn(),
}));

jest.mock("../../redux/ducks/orders", () => ({
  __esModule: true,
  selectOrderById: jest.fn(),
}));

const fakeOrder = {
  id: "123",
  name: "Test Order",
  date: "2023-06-01T12:00:00Z",
  price: 10000,
  status: "Pending",
  info: "Some info",
};

describe("OrderInfo", () => {
  test("should render order info", () => {
    window.history.pushState({}, "Test Title", "/test.html?id=123");

    (useAppSelector as jest.Mock).mockReturnValue(fakeOrder);

    (selectOrderById as jest.Mock).mockReturnValue(jest.fn());

    (useNavigate as jest.Mock).mockReturnValue(jest.fn());

    render(<OrderInfo />);

    expect(
      screen.getByText(/Test Order/i) &&
        screen.getByText(/10000/i) &&
        screen.getByText(/2023/i) &&
        screen.getByText(/Pending/i) &&
        screen.getByText(/Some info/i)
    ).toBeInTheDocument();
  });

  test('should render "Order not found" if order is not found', () => {
    window.history.pushState({}, "Test Title", "/?id=999");

    (useAppSelector as jest.Mock).mockReturnValue(undefined);
    render(<OrderInfo />);

    expect(screen.getByText(/Order not found/i)).toBeInTheDocument();
  });

  test("should navigate to root when 'Go back' button is clicked", () => {
    const mockNavigate = jest.fn();
    (useNavigate as jest.Mock).mockReturnValue(mockNavigate);

    render(<OrderInfo />);

    screen.getByText(/Go back/i).click();
    expect(mockNavigate).toHaveBeenCalledWith("/");
  });
});

describe("OrderInfo snapshot", () => {
  test("should match order info snapshot", () => {
    window.history.pushState({}, "", "/?id=123");

    (useNavigate as jest.Mock).mockReturnValue(jest.fn());
    (useAppSelector as jest.Mock).mockReturnValue(fakeOrder);

    const { asFragment } = render(<OrderInfo />);
    expect(asFragment()).toMatchSnapshot();
  });
});
