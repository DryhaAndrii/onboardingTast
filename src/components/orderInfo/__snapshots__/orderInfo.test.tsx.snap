// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderInfo snapshot should match a snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bXdtCk kTETQx"
  >
    <div>
      <p>
        Client: Test Order
      </p>
      <p>
        Date: 2023/06/01 (15:00)
      </p>
      <p>
        Price: 10000
      </p>
      <p>
        Status: Pending
      </p>
      <p>
        Info: Some info
      </p>
      <button>
        Go back
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`OrderInfo snapshot should match order info snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bXdtCk kTETQx"
  >
    <div>
      <p>
        Client: Test Order
      </p>
      <p>
        Date: 2023/06/01 (15:00)
      </p>
      <p>
        Price: 10000
      </p>
      <p>
        Status: Pending
      </p>
      <p>
        Info: Some info
      </p>
      <button>
        Go back
      </button>
    </div>
  </div>
</DocumentFragment>
`;
